### PR TITLE
Make `tcx.mk_const` more permissive wrt `kind` argument (`impl Into`)

### PR DESCRIPTION
**Pull request by WaffleLapkin**
r? @oli-obk you've asked for this >:)

### CODE REVIEW

File: compiler/rustc_middle/src/ty/consts.rs [hunk 1]
```diff
@@ -156,7 +150,7 @@ impl<'tcx> Const<'tcx> {
 
     pub fn from_scalar_int(tcx: TyCtxt<'tcx>, i: ScalarInt, ty: Ty<'tcx>) -> Self {
         let valtree = ty::ValTree::from_scalar_int(i);
-        Self::from_value(tcx, valtree, ty)
+        tcx.mk_const(valtree, ty)
     }
```
Comment (oli-obk) [inline compiler/rustc_middle/src/ty/consts.rs:154]: this method could also become redundant with an appropriate From impl

**Review by oli-obk — CHANGES_REQUESTED** [review]

**Comment by oli-obk** [comment]
@bors delegate+

r=me with perf clean and the review comment addressed

**Comment by oli-obk** [comment]
@rust-timer build 82877d5eeacaaf632388af232976d23d529ca7c2

**Comment by WaffleLapkin** [comment]
> regressions
> no action needed

??

Also, the regression itself looks super weird...

**Comment by oli-obk** [comment]
```
--------------------------------------------------------------------------------
Ir        
--------------------------------------------------------------------------------
6,087,305  PROGRAM TOTALS

--------------------------------------------------------------------------------
Ir          file:function
--------------------------------------------------------------------------------
-2,098,831  ???:<rustc_mir_build::build::Builder>::match_candidates
 2,020,333  ???:<rustc_mir_build::build::Builder>::simplify_candidate
-1,926,791  ???:<rustc_hir_typeck::fn_ctxt::FnCtxt>::require_type_is_sized
 1,720,536  ???:<rustc_trait_selection::traits::fulfill::FulfillmentContext as rustc_infer::traits::engine::TraitEngine>::register_bound
 1,490,315  ???:<rustc_mir_build::build::expr::as_place::PlaceBuilder>::field
 1,231,992  ???:<rustc_mir_build::build::expr::as_place::PlaceBuilder>::try_to_place
  -682,829  ???:<alloc::vec::Vec<rustc_middle::mir::syntax::ProjectionElem<rustc_middle::mir::Local, rustc_middle::ty::Ty>> as alloc::vec::spec_from_iter::SpecFromIter<rustc_middle::mir::syntax::ProjectionElem<rustc_middle::mir::Local, rustc_middle::ty::Ty>, core::iter::adapters::chain::Chain<core::iter::adapters::copied::Copied<core::slice::iter::Iter<rustc_middle::mir::syntax::ProjectionElem<rustc_middle::mir::Local, rustc_middle::ty::Ty>>>, core::array::iter::IntoIter<rustc_middle::mir::syntax::ProjectionElem<rustc_middle::mir::Local, rustc_middle::ty::Ty>, 1>>>>::from_iter
  -659,276  ???:rustc_mir_dataflow::drop_flag_effects::on_all_children_bits::is_terminal_path
   553,290  ???:do_rallocx
   529,097  ???:_rjem_je_arena_ralloc
   501,018  ???:<rustc_middle::ty::FieldDef>::ty
   476,288  ???:<alloc::raw_vec::RawVec<rustc_middle::mir::syntax::ProjectionElem<rustc_middle::mir::Local, rustc_middle::ty::Ty, rustc_middle::ty::Ty>>>::reserve_for_push
   375,478  ???:<rustc_middle::ty::subst::SubstFolder as rustc_middle::ty::fold::FallibleTypeFolder>::try_fold_ty
   353,556  ???:_rjem_je_arena_ralloc_no_move
```

I... have no idea ^^ I remember seeing `field` and `try_to_place` in other regressions on html5ever, so it's likely inliner noise.

**Pull request merged**